[#100412960] Terrraform deployment for MicroBOSH

saliceti opened the pull request:
## What

We have agreed to start our CF test by using the microBOSH and scaling to full BOSH only if needed. This story is to create the first infrastructure needed for the CF build out by creating microBOSH instances in AWS + GCE.

Please refer to the README for implementation details.
## How to review

Create the microBOSH VM on each cloud by following the instructions in the README. The success criteria is having the bastion created, the bosh-init script running and creating the BOSH VM until getting messages such as:

```
Waiting for instance 'bosh/0' to be running...
Finished (00:00:12)
```

**Please note:**
- it takes about 15 min for each platform
- there is an intermittent issue on GCE whereby the BOSH VM does not always listen on SSH port and leaves the bosh-init script hanging. The workaround is to reset the VM via the GCE console and the bosh-init script should continue
## Who should review

Anyone but @jimconner and @saliceti

dcarley committed:
Copy stub files from tsuru-terraform

Variables that we will need for this project. Excluding things that we won't
need (GCS) and things that we don't need yet (network subnets).
gce: Copy provider.tf from tsuru-terraform to configure the provider for GCE.
gce: Add globals.tf with "env" variable. Pattern copied from tsuru-terraform to allow independent environments.

dcarley committed:
Adding .gitignore

dcarley committed:
gce: Add bastion network

This network will host the NAT and microbosh instances.

dcarley committed:
gce: Add SSH firewall rule

Which will allow us to SSH to the bastion host that we will later create.

dcarley committed:
gce: Add bastion host

Which will run bosh-init.

dhilton committed:
gce: Create template manifest for microbosh to use

dhilton committed:
GCE: Added static IP for microbosh to use once provisioned

dhilton committed:
gce: Output the static IP for microbosh.

dhilton committed:
Added new gce_region_zone

Due to microbosh init needing a precise avalilablity zone into which to
provision, we created a new variable to allow this to change.

dhilton committed:
Specify ssh_user for bosh-init

dhilton committed:
Added provision.sh to bootstrap microbosh install

Have created a bash script to boot the install of bosh-init and
configuration of the manifest.yml ready for the microbosh install.

Commit:
Allocate static IP for bastion host

We need to have the firewall rules in place before
creation of the bastion host, so we have to assign
a static IP in advance - otherwise terraform will
attempt to complete provisioning on the bastion hosts
before it creates the relevant firewall rules to
allow provisioning to succeed.

Commit:
Use insecure-deployer ssh key instead of id_rsa

Ensure that the ubuntu@bastion host is using the
insecure-deployer ssh key as bosh-init does not
seem to pick up any keys apart from id_rsa

Commit:
Switch to release 168

Relase 194 wanted some sections of config which were not present
in our template.

Commit:
First draft of AWS support

saliceti committed:
Moved provisioning to a Makefile

Contains commands to apply, provision, delete bosh, destroy, terraform, ssh to bastion, for both AWS and GCE.

Commit:
Switch to bosh v190 as it works

Problems rendering manifests with v194,v195 and v196
Switching to v190 resolves this problem.

Commit:
Bumped bosh init version to 0.0.72

Commit:
Update readme with usage instructions

saliceti committed:
Removed bastion elastic ip in GCE

saliceti committed:
Changed name of manifest to standardise AWS and GCE

keymon committed:
Fix minor typos and adjusting some variables

Fix typos using the ssh keys, tag names, right network ranges

keymon committed:
Parametrise region and security groups on aws

keymon committed:
Adjust the firewall to limit access from office

Only office IPs can connect to the microbosh host. 
The elastic ip must be part of vpc.

keymon committed:
Remove unnecesary associate_public_ip_address

It is true by default.

keymon commented:
I tested in from my computer:

1. Install bosh_cli (I user rvm): `rvm ruby-2.2.2@cloudfoundry --create && gem install bosh_cli --no-document`
2. Try to login and get status (the password from the manifest is currently 'admin'): `yes admin | bosh target $(terraform output -state=hectorcf.tfstate bosh_ip) && bosh status`
 
Works on AWS! :+1: 

```
$ cd aws
$ yes admin| bosh target $(terraform output -state=hectorcf.tfstate bosh_ip) && bosh status
Target set to `my-bosh'
Config
             /Users/keymon/.bosh_config

Director
  Name       my-bosh
  URL        https://52.19.119.128:25555
  Version    1.3029.0 (00000000)
  User       admin
  UUID       b86cd765-7b10-4e06-a7d1-dc3ea45becc7
  CPI        cpi
  dns        disabled
  compiled_package_cache disabled
  snapshots  disabled

Deployment
  not set

```

keymon commented:
And GCE works fine as well (although I had to update the firewall to include the office, commit on its way):

```
$ yes admin | bosh target $(terraform output -state=hectorcf.tfstate bosh_ip) && bosh status
Target set to `micro-google'
Your username: Enter password: stty: stdin isn't a terminal
stty: stdin isn't a terminal
stty: stdin isn't a terminal

Logged in as `admin'
Config
             /Users/keymon/.bosh_config

Director
  Name       micro-google
  URL        https://104.155.73.241:25555
  Version    1.3029.0 (00000000)
  User       admin
  UUID       d877d2ca-6f0a-4882-a566-2cde57d27c77
  CPI        cpi
  dns        enabled (domain_name: microbosh)
  compiled_package_cache disabled
  snapshots  disabled

Deployment
  not set
```

keymon committed:
Allow office IPs to access public bosh on GCE

We need to add the office addresses in the GCE firewall for the public
bosh IP in order to use the bosh client from the local computer.

keymon committed:
Suggest gcloud command to reset vm in the README

keymon commented:
All good!!! Nice work

Merging!

annashipman commented:
Nice work all. Thanks @keymon for a really thorough review and for picking up commits that could be split as well, really appreciate that as it makes the history much clearer to follow later. Thanks all!
